src/goto-harness/function_call_harness_generator.cpp

```diff
@@ -125,55 +196,38 @@ void function_call_harness_generatort::generate(
   p_impl->generate(goto_model, harness_function_name);
 }
 
-void function_call_harness_generatort::implt::
-  setup_parameters_and_call_entry_function(code_blockt &function_body)
-{
-  const auto &function_to_call = lookup_function_to_call();
-  const auto &function_type = to_code_type(function_to_call.type);
-  const auto &parameters = function_type.parameters();
-
-  code_function_callt::operandst arguments{};
-
-  auto allocate_objects = allocate_objectst{function_to_call.mode,
-                                            function_to_call.location,
-                                            "__goto_harness",
-                                            *symbol_table};
-  for(const auto &parameter : parameters)
-  {
-    auto argument = allocate_objects.allocate_automatic_local_object(
-      parameter.type(), parameter.get_base_name());
-    arguments.push_back(argument);
-  }
-  allocate_objects.declare_created_symbols(function_body);
-  for(auto const &argument : arguments)
-  {
-    generate_initialisation_code_for(function_body, argument);
-  }
-  code_function_callt function_call{function_to_call.symbol_expr(),
-                                    std::move(arguments)};
-  function_call.add_source_location() = function_to_call.location;
-
-  function_body.add(std::move(function_call));
-}
-
 void function_call_harness_generatort::implt::generate(
   goto_modelt &goto_model,
   const irep_idt &harness_function_name)
 {
   symbol_table = &goto_model.symbol_table;
   goto_functions = &goto_model.goto_functions;
-  const auto &function_to_call = lookup_function_to_call();
-  recursive_initialization_config.mode = function_to_call.mode;
-  recursive_initialization = util_make_unique<recursive_initializationt>(
-    recursive_initialization_config, goto_model);
   this->harness_function_name = harness_function_name;
+  const auto &function_to_call = lookup_function_to_call();
   ensure_harness_does_not_already_exist();
 
   // create body for the function
   code_blockt function_body{};
+  auto const arguments = declare_arguments(function_body);
+
+  // configure and create recursive initialisation object
+  recursive_initialization_config.mode = function_to_call.mode;
+  recursive_initialization_config.pointers_to_treat_as_arrays =
+    function_arguments_to_treat_as_arrays;
+  recursive_initialization_config.array_name_to_associated_array_size_variable =
+    function_argument_to_associated_array_size;
+  std::transform(
+    function_argument_to_associated_array_size.begin(),
+    function_argument_to_associated_array_size.end(),
+    std::inserter(
+      recursive_initialization_config.variables_that_hold_array_sizes,
+      recursive_initialization_config.variables_that_hold_array_sizes.end()),
+    [](const std::pair<irep_idt, irep_idt> &p) { return p.second; });
```

Comment:
🤔 In trivial cases like this a for-loop might be clearer:
```c++
for (const auto &pair : function_argument_to_associated_array_size)
  recursive_initialization_config.variables_that_hold_array_sizes(pair.second);
```
These names are insane BTW.